contracts/DisputableDelay.sol

```diff
@@ -0,0 +1,264 @@
+        Paused,              // A delayed script that is being challenged by Agreements
+        Cancelled,           // A delayed script that has been cancelled by Agreements
+        Executed             // A delayed script that has been executed
+    }
+
+    struct DelayedScript {
+        uint64 executionFromTime;
+        uint64 pausedAt;
+        DelayedScriptStatus delayedScriptStatus;
+        bytes evmCallScript;
+        uint256 actionId;
+    }
+
+    uint64 public executionDelay;
+    uint256 public delayedScriptsNewIndex = 0;
+    mapping (uint256 => DelayedScript) public delayedScripts;
+
+    event ExecutionDelaySet(uint64 indexed executionDelay);
+    event DelayedScriptStored(uint256 indexed delayedScriptId, uint256 indexed actionId, bytes evmScript);
+    event ExecutionPaused(uint256 indexed delayedScriptId, uint256 indexed actionId);
+    event ExecutionResumed(uint256 indexed delayedScriptId, uint256 indexed actionId);
+    event ExecutionCancelled(uint256 indexed delayedScriptId, uint256 indexed actionId);
+    event AgreementActionClosed(uint256 indexed delayedScriptId, uint256 indexed actionId);
+    event ExecutedScript(uint256 indexed delayedScriptId, uint256 indexed actionId);
+
+    /**
+    * @notice Initialize the Delay app
+    * @param _executionDelay The delay in seconds a user will have to wait before executing a script
+    */
+    function initialize(uint64 _executionDelay) external onlyInit {
+        initialized();
+        executionDelay = _executionDelay;
+    }
+
+    /**
+    * @notice Set the execution delay to `_executionDelay`
+    * @param _executionDelay The new execution delay
+    */
+    function setExecutionDelay(uint64 _executionDelay) external auth(SET_DELAY_ROLE) {
+        executionDelay = _executionDelay;
+
+        emit ExecutionDelaySet(executionDelay);
+    }
+
+    /**
+    * @dev IDisputable interface conformance
+    */
+    function getDisputableAction(uint256 _delayedScriptId)
+        external view returns (uint64 endDate, bool challenged, bool finished)
+    {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        DelayedScriptStatus delayedScriptStatus = delayedScript.delayedScriptStatus;
+
+        endDate = delayedScript.executionFromTime;
+        challenged = delayedScriptStatus == DelayedScriptStatus.Paused;
+        finished = delayedScriptStatus == DelayedScriptStatus.Cancelled
+            || delayedScriptStatus == DelayedScriptStatus.Executed;
+    }
+
+    /**
+    * @dev IDisputable interface conformance
+    */
+    function canChallenge(uint256 _delayedScriptId) external view returns (bool) {
+        return _canPause(delayedScripts[_delayedScriptId]);
+    }
+
+    /**
+    * @dev IDisputable interface conformance
+    */
+    function canClose(uint256 _delayedScriptId) external view returns (bool) {
+        return _canClose(delayedScripts[_delayedScriptId]);
```

Comment:
⚠️ Should we allow to close a delayable even if it cannot be executed for now? (especially the case when the `executeFromTime` date has not been reached yet)